### PR TITLE
Implement termset pairwise similarity with weights and negated phenotypes

### PR DESCRIPTION
**Pull request by justaddcoffee**
Experimental implementation of termset_pairwise_similarity_weighted_negated(), which 
is similar to termset_pairwise_similarity, but it accepts weights for each term,
such that the similarity score is weighted by the term's weight.

Also, this function accepts a set of negated terms, which are terms that have been excluded
by the clinician. If negated, a matching term will reduce the similarity

This is used for example in N3C work, in which we have a Spark DF of "profiles":
profile_id  term_id  weight  negated
1           HP:1234  0.5     false
1           HP:5678  0.2     false
1           HP:9012  0.3     true
2           HP:6789  0.3     false
2           HP:3456  0.7     false

and a Spark DF of Patient phenotypes:
patient_id  term_id  negated
1           HP:1234  false
1           HP:5678  false
1           HP:9012  false
2           HP:6789  false
2           HP:3456  true
...

   Arguments

        subject_dat: &Vec<(HPO term: String, weight: f64, negated: booll)>,
        object_dat: &Vec<(HPO term: String, weight: f64, negated: booll)>,

### CODE REVIEW

**Comment by justaddcoffee** [comment]
@hrshdhgd what do you think of this PR

also, does [this](https://github.com/monarch-initiative/semsimian/blob/0ee8256d43bdc846875eb9d6c2d2d60e78743900/src/lib.rs#L908) code do everything we need to do to make this `termset_pairwise_similarity_weighted_negated()` available in Python? 

```    
fn termset_pairwise_similarity_weighted_negated(
        &mut self,
        subject_dat: Vec<(TermID, f64, bool)>,
        object_dat: Vec<(TermID, f64, bool)>
     ) -> PyResult<f64> {
        self.ss.update_closure_and_ic_map();
        return Ok(self.ss.termset_pairwise_similarity_weighted_negated(&subject_dat, &object_dat));
    }
```

File: src/lib.rs [hunk 1]
```diff
@@ -437,6 +443,87 @@ impl RustSemsimian {
...
+            for (key2, value2) in value1.iter() {
+                all_by_all_object_perspective
+                    .entry(key2.to_owned())
+                    .or_insert_with(HashMap::new)
+                    .insert(key1.to_owned(), value2.to_owned());
+            }
+        }
+        // return self.termset_comparison(&subject_terms, &object_terms).unwrap();
+        let subject_to_object_average_resnik_sim: f64 = calculate_weighted_term_pairwise_information_content(
+            &self.closure_map,
```
Comment (hrshdhgd) [inline src/lib.rs:502]: You can just pass &self here instead of 2 params.

File: src/lib.rs [hunk 2]
```diff
@@ -437,6 +443,87 @@ impl RustSemsimian {
...
+            &self.ic_map,
+            &subject_dat,
+            &object_dat,
+            &self.predicates,
+        );
+
+
+        let object_to_subject_average_resnik_sim: f64 = calculate_weighted_term_pairwise_information_content(
+            &self.closure_map,
+            &self.ic_map,
```
Comment (hrshdhgd) [inline src/lib.rs:512]: Same as above.

**Comment by hrshdhgd** [comment]
> @hrshdhgd what do you think of this PR
> 
> also, does [this](https://github.com/monarch-initiative/semsimian/blob/0ee8256d43bdc846875eb9d6c2d2d60e78743900/src/lib.rs#L908) code do everything we need to do to make this `termset_pairwise_similarity_weighted_negated()` available in Python?
> 
> ```rust
> fn termset_pairwise_similarity_weighted_negated(
>         &mut self,
>         subject_dat: Vec<(TermID, f64, bool)>,
>         object_dat: Vec<(TermID, f64, bool)>
>      ) -> PyResult<f64> {
>         self.ss.update_closure_and_ic_map();
>         return Ok(self.ss.termset_pairwise_similarity_weighted_negated(&subject_dat, &object_dat));
> }
> ```

A python test will confirm if this is enough but I think this should do it.

**Comment by hrshdhgd** [comment]
I see there is overlap in code with the existing `fn termset_pairwise_similarity()`. Could we just reuse this function with an additional param (weighted: bool) which decides which way to go? I think it'll keep the code minimal .

Thoughts?

**Review by hrshdhgd — CHANGES_REQUESTED** [review]
- [ ] Add python test
- [x] Rewire `termset_pairwise_similarity()` to be reusable
- [x] Reduce redundant params

File: src/lib.rs [hunk 3]
```diff
@@ -437,6 +454,74 @@ impl RustSemsimian {
...
+        // # Arguments
+        //        subject_dat: tuples of terms for termset 1 (term_id, weight, negated)
+        //        object_dat: tuples of terms for termset 2 (term_id, weight, negated)
+
+        let subject_terms = subject_dat.iter().map(|(term, _, _)| term.clone()).collect::<HashSet<TermID>>();
+        let object_terms = object_dat.iter().map(|(term, _, _)| term.clone()).collect::<HashSet<TermID>>();
+
+        let all_by_all: SimilarityMap;
+        let all_by_all_object_perspective: SimilarityMap;
+        (all_by_all, all_by_all_object_perspective) = self.get_both_all_by_all_objects(&subject_terms, &object_terms);
```
Comment (hrshdhgd) [inline src/lib.rs:500]: I made the changes I suggested, but the variables `all_by_all` and `all_by_all_object_perspective` don't seem to be used in this function. Is something pending?

Comment (justaddcoffee) [reply src/lib.rs]: nope, just didn't notice this - I've removed these unused variables

File: src/lib.rs [hunk 4]
```diff
@@ -1290,6 +1359,124 @@ mod tests {
         assert_eq!(tsps.best_score, 5.8496657269155685);
     }
 
+    #[rstest(
```
Comment (hrshdhgd) [inline src/lib.rs:1362]: Can this be moved to `test_utils.rs` along with other constants we use in tests ?

Comment (justaddcoffee) [reply src/lib.rs:1362]: @hrshdhgd move what specifically? these are the parameters for this parameterized test, not sure how/why exactly you'd move them to a different file

Comment (hrshdhgd) [reply src/lib.rs:1362]: > (discuss at hackathon today?)

Yes.

> not sure how/why exactly you'd move them to a different file

The lib.rs file is getting too large (already ~2000 lines). It'll be hard to maintain if it explodes (by this I mean scrolling up and down too often which we are already kind of doing). We'll need a new PR to break out code into different modules once search is dealt with.

**Comment by justaddcoffee** [comment]
merging per convo with @hrshdhgd today

**Pull request merged**